### PR TITLE
Find last checkpoints on restart

### LINKED ISSUE #14466: Last saved checkpoint not resumed when using trainer.fit(..., ckpt_path='last')
## 🐛 Bug

Last saved checkpoint (by setting save_last=True in ModelCheckpoint callback) is not automatically resumed when specifying ckpt_path='last' argument in trainer.fit method. 

From the related code (https://github.com/Lightning-AI/lightning/blob/1.7.3/src/pytorch_lightning/trainer/trainer.py#L1419), I see that the last model path is taken from the "last_model_path" attribute of the ModelCheckpoint callback object, but its value is always an empty string at the moment of resuming the latest checkpoint.

One workaround is to manually set this attribute just after creating the callback, but I don't know if this was the intended behaviour or if there is another (and possibly better) way to do all this.

### To Reproduce

Run this code a first time and stop it after some epochs have been completed. Then run again. The second run does not resume the latest saved checkpoint but starts the training again from the first epoch.

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset
from time import sleep
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import ModelCheckpoint


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def on_validation_epoch_end(self) -> None:
        sleep(1)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():

    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        callbacks=ModelCheckpoint(monitor='valid_loss', mode='min', filename='bestLoss', save_last=True),
        accelerator='cpu',
        logger=None,
        default_root_dir=os.getcwd(),
        max_epochs=1000)

    # fit
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data, ckpt_path='last')


if __name__ == "__main__":
    run()

```

### Expected behavior

Resume training from the last saved checkpoint when using ckpt_path="last" option in trainer.fit

### Environment
<details>
  <summary>Details</summary>

        * CUDA:
                - GPU:
                        - Quadro T1000
                - available:         True
                - version:           10.2
        * Lightning:
                - pytorch-lightning: 1.7.3
                - torch:             1.12.1
                - torchmetrics:      0.9.3
        * Packages:
                - absl-py:           1.2.0
                - aiohttp:           3.8.1
                - aiosignal:         1.2.0
                - async-timeout:     4.0.2
                - attrs:             22.1.0
                - cachetools:        5.2.0
                - certifi:           2022.6.15
                - charset-normalizer: 2.1.1
                - frozenlist:        1.3.1
                - fsspec:            2022.7.1
                - google-auth:       2.11.0
                - google-auth-oauthlib: 0.4.6
                - grpcio:            1.48.0
                - idna:              3.3
                - importlib-metadata: 4.12.0
                - markdown:          3.4.1
                - markupsafe:        2.1.1
                - multidict:         6.0.2
                - numpy:             1.23.2
                - oauthlib:          3.2.0
                - packaging:         21.3
                - pip:               22.2.2
                - protobuf:          3.19.4
                - pyasn1:            0.4.8
                - pyasn1-modules:    0.2.8
                - pydeprecate:       0.3.2
                - pyparsing:         3.0.9
                - pytorch-lightning: 1.7.3
                - pyyaml:            6.0
                - requests:          2.28.1
                - requests-oauthlib: 1.3.1
                - rsa:               4.9
                - setuptools:        63.3.0
                - six:               1.16.0
                - tensorboard:       2.10.0
                - tensorboard-data-server: 0.6.1
                - tensorboard-plugin-wit: 1.8.1
                - torch:             1.12.1
                - torchmetrics:      0.9.3
                - tqdm:              4.64.0
                - typing-extensions: 4.3.0
                - urllib3:           1.26.12
                - werkzeug:          2.2.2
                - wheel:             0.37.1
                - yarl:              1.8.1
                - zipp:              3.8.1
        * System:
                - OS:                Linux
                - architecture:
                        - 64bit
                        - ELF
                - processor:         x86_64
                - python:            3.8.10
                - version:           #49~20.04.1-Ubuntu SMP Thu Aug 4 19:15:44 UTC 2022
</details>

### Additional context

cc @justusschock @awaelchli @ninginthecloud @rohitgr7 @otaj @akihironitta

### PR DESCRIPTION
**Pull request by otaj**
## What does this PR do?

Fixes #14466

Although #12816 enabled support for passing `"last"` as an argument for reloading checkpoint, reloading happened only for FT checkpoints and checkpoints that were created within the same run.

This PR adds support for reloading last checkpoint on restart of the run.

Note, that this makes `"best"` and `"last"` keywords divergent in their behavior - `"best"` doesn't reload the best available checkpoint when starting a new run for training.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @awaelchli @ananthsub @ninginthecloud @rohitgr7 @otaj @carmocca @jjenniferdai @akihironitta

### CODE REVIEW

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -583,6 +583,8 @@ def __resolve_ckpt_dir(self, trainer: "pl.Trainer") -> None:
         """
         if self.dirpath is not None:
             # short circuit if dirpath was passed to ModelCheckpoint
+            if broadcast:
+                self.dirpath = trainer.strategy.broadcast(self.dirpath)
```
Comment (awaelchli) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:587]: This is too dangerous for my taste. It can lead to unwanted hangs when dirpath is None on one rank and not None on another. Even if it's not the case today, it's annoying to debug if you run into it. Our best practice in Lightning is to never run collectives under a conditional unless you can guarantee that every rank enters it. 

I suggest to keep the broadcasting outside of `__resolve_ckpt_dir`, i.e., wherever it is needed, call

```
dir = __resolve_ckpt_dir()
dir = broadcast(dir)
```
and otherwise not. 

What do you think?

Comment (otaj) [reply src/pytorch_lightning/callbacks/model_checkpoint.py:587]: That absolutely makes sense. Will make the required changes

Comment (otaj) [reply src/pytorch_lightning/callbacks/model_checkpoint.py]: Done. Thank you!

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -255,7 +255,9 @@ def state_key(self) -> str:
         )
 
     def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: str) -> None:
-        self.__resolve_ckpt_dir(trainer)
+        dirpath = self.__resolve_ckpt_dir(trainer)
+        dirpath = trainer.strategy.broadcast(dirpath)
+        self.dirpath = dirpath
         assert self.dirpath is not None
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:261]: Not necessary now

```suggestion
```

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -600,9 +602,14 @@ def __resolve_ckpt_dir(self, trainer: "pl.Trainer") -> None:
...
 
-        ckpt_path = trainer.strategy.broadcast(ckpt_path)
+        return ckpt_path
 
-        self.dirpath = ckpt_path
+    def _find_last_checkpoints(self, trainer: "pl.Trainer") -> List[str]:
+        # find all checkpoints in the folder
+        ckpt_path = self.__resolve_ckpt_dir(trainer)
+        if self._fs.exists(ckpt_path):
+            return [os.path.normpath(p) for p in self._fs.ls(ckpt_path, detail=False) if self.CHECKPOINT_NAME_LAST in p]
```
Comment (carmocca) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:611]: Why do you normpath?

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 4]
```diff
@@ -161,9 +161,10 @@ def _set_ckpt_path(
             ckpt_path = getattr(self.trainer.checkpoint_callback, "best_model_path", None)
 
         if ckpt_path == "last":
-            candidates = [getattr(ft, "ckpt_path", None) for ft in ft_checkpoints] + [
-                getattr(cb, "last_model_path", None) for cb in self.trainer.checkpoint_callbacks
-            ]
+            candidates = [getattr(ft, "ckpt_path", None) for ft in ft_checkpoints]
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:164]: Should all of these be sets?

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 5]
```diff
@@ -161,9 +161,10 @@ def _set_ckpt_path(
...
+            for callback in self.trainer.checkpoint_callbacks:
+                if hasattr(callback, "_find_last_checkpoints"):
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:166]: For customization, we'd want to have this as something the users can subclass and override with some guarantees. I guess this is why you chose to use `hasattr`.

But I would check `isinstance(ModelCheckpoint)` instead in this PR

File: src/pytorch_lightning/callbacks/model_checkpoint.py (hunk 3 above)
Comment (otaj) [reply src/pytorch_lightning/callbacks/model_checkpoint.py:612]: Because otherwise tests fail on Windows, because `self._fs.ls` uses `posixpath.separator` on local filesystem regardless of platform

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 4 above)
Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py]: Good idea, done

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 5 above)
Comment (otaj) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py]: Yeah, that makes sense since we probably cannot expect people to implement this in the same way as us if they subclass a checkpoint. Done

**Review by carmocca — COMMENTED** [review]
LGTM. Waiting on a test though!

**Comment by awaelchli** [comment]
I found some tests related to this in `test_trainer.py`, `test_checkpoint_path_input_last` etc.. Perhaps one could modify one of these or follow a similar pattern as these tests.

**Comment by otaj** [comment]
@awaelchli, @carmocca there is a test now. Btw, writing it helped me uncover a small bug with the code, see d3a3b35

File: tests/tests_pytorch/trainer/test_trainer.py [hunk 6]
```diff
@@ -776,6 +776,40 @@ def test_checkpoint_path_input_last(tmpdir, ckpt_path, save_last, fn):
...
 
 
+def test_checkpoint_find_last(tmpdir):
+    """Test that the last checkpoint is found correctly."""
+    model = BoringModel()
+    mc = ModelCheckpoint(save_last=True)
+    trainer = Trainer(
+        max_epochs=1,
+        limit_train_batches=1,
+        limit_val_batches=1,
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/test_trainer.py:786]: Probably does not need validation

```suggestion
        limit_val_batches=0,
```

File: tests/tests_pytorch/trainer/test_trainer.py [hunk 7]
```diff
@@ -776,6 +776,40 @@ def test_checkpoint_path_input_last(tmpdir, ckpt_path, save_last, fn):
...
+    )
+    assert trainer.ckpt_path is None
+    trainer.fit(model)
+
+    model = BoringModel()
+    mc = ModelCheckpoint()
+    trainer = Trainer(
+        max_epochs=1,
+        limit_train_batches=1,
+        limit_val_batches=1,
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/test_trainer.py:801]: Same comment

File: tests/tests_pytorch/trainer/test_trainer.py [hunk 8]
```diff
@@ -776,6 +776,40 @@ def test_checkpoint_path_input_last(tmpdir, ckpt_path, save_last, fn):
...
+        enable_model_summary=False,
+        enable_progress_bar=False,
+        logger=False,
+        default_root_dir=tmpdir,
+        callbacks=[mc],
+    )
+    assert trainer.ckpt_path is None
+    trainer.fit(model)
+
+    model = BoringModel()
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/test_trainer.py:796]: You probably can get away with reusing the previous instance

```suggestion
```

File: tests/tests_pytorch/trainer/test_trainer.py [hunk 9]
```diff
@@ -776,6 +776,40 @@ def test_checkpoint_path_input_last(tmpdir, ckpt_path, save_last, fn):
...
+    mc = ModelCheckpoint()
+    trainer = Trainer(
+        max_epochs=1,
+        limit_train_batches=1,
+        limit_val_batches=1,
+        enable_model_summary=False,
+        enable_progress_bar=False,
+        logger=False,
+        default_root_dir=tmpdir,
+        callbacks=[mc],
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/test_trainer.py:806]: This does not require pasing the checkpoint callback again, no?

File: tests/tests_pytorch/trainer/test_trainer.py (hunk 8 above)
Comment (otaj) [reply tests/tests_pytorch/trainer/test_trainer.py:796]: I want to be sure we're simulating restarting the run as closely as possible.

File: tests/tests_pytorch/trainer/test_trainer.py (hunk 9 above)
Comment (otaj) [reply tests/tests_pytorch/trainer/test_trainer.py:806]: It does - we are finding last checkpoints as a method on `ModelCheckpoint` instance

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
Great!
